contributors/devel/sig-instrumentation/migration-to-structured-logging.md

```diff
@@ -0,0 +1,329 @@
+
+
+Simplified mapping between functions:
+* `klog.Infof`, `klog.Info`, `klog.Infoln`, `klog.InfoDepth` -> `klog.InfoS`
+* `klog.V(N).Infof`, `klog.V(N).Info`, `klog.V(N).Infoln` -> `klog.V(N).InfoS`
+* `klog.Warning`, `klog.Warningf`, `klog.Warningln`, `klog.WarningDepth` -> `klog.InfoS`
+* `klog.Error`, `klog.Errorf`, `klog.Errorln`, `klog.ErrorDepth` -> `klog.ErrorS`
+* `klog.Fatal`, `klog.Fatalf`, `klog.Fatalln`, `klog.FatalDepth` -> `klog.ErrorS`
+
+### Removing Depth
+
+Functions with depth (`klog.InfoDepth`, `klog.WarningDepth`, `klog.ErrorDepth`, `klog.FatalDepth`) are used to indicate
+that source of log (added as metadata in log) is different then invocation of logging library. This is usually used when
+implementing logging util functions. In those situations util functions should be removed and `klog` call moved up.
+
+For example
+```go
+func Handle(w http.ReponseWriter, r *http.Request) {
+    logHTTPRequest(r)
+    handle(w, r)
+}
+func logHTTPRequest(r *http.Request) {
+    klog.InfoDepth(1, "Got request")
+}
+```
+should be replaced with
+```go
+func Handle(w http.ReponseWriter, r *http.Request) {
+    klog.Info("Got request")
+    handle(w, r)
+}
+```
+
+### Using ErrorS
+
+With `klog` structured logging borrowing interface from [logr] it also inherits it's differences in semantic of error
+function. This is visible in `klog.ErrorS` which requires passing an `error` as first argument. This is because in
+[logr] Error is used to inform users about errors received from subordinate function calls.
+This means that `klog.ErrorS` should only be used when there is a obvious Golang `error` object available.
+
+Creating error just to call `klog.ErrorS` or passing `nil` is discouraged, please consider using `klog.InfoS` instead.
+
+### Replacing Fatal calls
+
+Usage of Fatal should be discouraged and it's not available in new functions . Instead of depending on logger to exist
+process you should consider returning error and ensuring it is properly handled.
+
+## Remove string formatting from log message
+
+With structured logging, log messages are no longer formatted, leaving argument marshalling up to logging client
+implementation. This allows messages to be a static description of event.
+
+All string formatting (`%d`, `%v`, `%w`, `%s`) should be removed and message simplified.
+Describing arguments in log messages is no longer needed and should be removed leaving only description of what happen.
+
+For example
+```go
+klog.Infof("delete object %s with propagation policy %s", ...)
+```
+should be changed to
+```go
+klog.Infof("delete object", ...)
+```
+
+Some logs are constructed solely from string formats. In those cases a message needs to be derived from context of
+log call.
+
+For example http access logs
+```go
+func LogHTTP(r *http.Request) {
+   klog.Infof("%s %s: (%v) %v%v%v [%s %s]", ...)
+}
+```
+should be changed to
+```go
+func LogHTTP(r *http.Request) {
+   klog.InfoS("HTTP", ...)
+}
+```
+
+### Name arguments
```

Comment:
I guess that even a short list of standard argument names like "status" could be helpful and would promote consistency.

Reply:
Do you have in mind listing arguments other then kubernetes kinds (those should be obvious)? I wanted to have this list evolve naturally during migration, instead of trying to force set of keys at this point. 

With log calls migrated it will be really easy to track global list of keys and propose refactors on whole repo.

Reply:
Yes, I was thinking about arguments like "status", "reason", "errorCode", "result" etc.
I am ok with creating such list latter on but I think that it would be good to have it somewhere in the documentation so that for common arguments we could have a proper names just from the start.